File: shells/android-template/app/src/main/java/dev/wefter/bridge/plugins/PingTestPlugin.kt
package dev.wefter.bridge

import android.content.Context
import android.os.Handler
import android.os.Looper
import org.json.JSONObject

private const val TICK_INTERVAL_MS = 2000L

class PingTestPlugin(context: Context, dispatcher: BridgeDispatcher) : WefterPlugin(context, dispatcher) {
    private val handler = Handler(Looper.getMainLooper())
    private var tickCount = 0

    private val tickRunnable =
            object : Runnable {
                override fun run() {
                    tickCount++
                    emit("tick", JSONObject().apply { put("count", tickCount) })
                    handler.postDelayed(this, TICK_INTERVAL_MS)
                }
            }

    init {
        handler.post(tickRunnable)
    }

    @WefterMethod
    fun ping(payload: JSONObject, callback: (Result<Any>) -> Unit) {
        val result = JSONObject().apply { put("pong", true) }
        resolve(callback, result)
    }
}

File: plugins/ping-test/ios/PingTestPlugin.swift
import Foundation
import UIKit

final class PingTestPlugin: WefterPlugin {
    private static let tickIntervalSeconds: TimeInterval = 2
    private var timer: Timer?
    private var tickCount = 0

    required init(dispatcher: BridgeDispatcher, viewController: UIViewController?) {
        super.init(dispatcher: dispatcher, viewController: viewController)
        startTicking()
    }

    private func startTicking() {
        let timer = Timer.scheduledTimer(withTimeInterval: Self.tickIntervalSeconds, repeats: true) { [weak self] _ in
            guard let self else { return }
            self.tickCount += 1
            self.emit("tick", ["count": self.tickCount])
        }
        RunLoop.main.add(timer, forMode: .common)
        self.timer = timer
    }

    // @WefterMethod
    func ping(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        resolve(callback, data: ["pong": true])
    }
}
